File: 1863.sum-of-all-subset-xor-totals.py
#
# @lc app=leetcode id=1863 lang=python3
#
# [1863] Sum of All Subset XOR Totals
#

# @lc code=start
import functools
class Solution:
    memo = {}

    def subsetXORSum(self, nums: List[int]) -> int:
        sum = 0
        self.memo = {} # WOW -- this needs to be reset or it carries between cases
        subsets = self.getSubsets(nums)
        for sub in subsets:
            sum += self.XORSum(sub)

        return sum

    def getSubsets(self, nums, keys=None):
        if not keys:
            keys = list(range(len(nums)))

        memoKey = str(keys)
        if len(nums) == 0 or memoKey in self.memo:
            return []

        sets = [nums]
        self.memo[memoKey] = True

        for i in range(len(nums)):
            newSet = nums[:]
            newSet.pop(i)
            newKeys = keys[:]
            newKeys.pop(i)

            sets += self.getSubsets(newSet, newKeys)
        return sets

    def XORSum(self, nums):
        return functools.reduce(lambda a,v: a^v, nums)
# @lc code=end

File: js/1863.sum-of-all-subset-xor-totals.js
/*
 * @lc app=leetcode id=1863 lang=javascript
 *
 * [1863] Sum of All Subset XOR Totals
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {number}
 */
var subsetXORSum = function(nums, start=0) {
  let subsets = getAllSubsets(nums)
  return subsets.reduce((a,v)=>{
    return a + xorTotal(v)
  },0)
};

let xorTotal = (arr) => {
  if (arr.length === 0) return 0
  else if (arr.length === 1) return arr[0]
  return arr.slice(1).reduce((a,v)=>a^v,arr[0])
}

// let getSubsets = (nums, set=[])=>{
//   if (nums.length === 1)
//     return [[nums[0]]]

//   let ubersets = []
//   for (let i=1; i<nums.length; i++) {
//     let subsets = getSubsets(nums.slice(i))
//     console.log(subsets)
//     subsets = subsets.map(arr=>[nums[0], ...arr])
//     ubersets = [...ubersets, ...subsets]
//   }
//   return ubersets
// }

const getAllSubsets =
      theArray => theArray.reduce(
        (subsets, value) => subsets.concat(
         subsets.map(set => [value,...set])
        ),
        [[]]
      );
// @lc code=end

